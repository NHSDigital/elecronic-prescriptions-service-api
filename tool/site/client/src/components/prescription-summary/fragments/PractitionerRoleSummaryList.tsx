import {HealthcareService, Identifier, Location, Organization, Practitioner, PractitionerRole} from "fhir/r4"
import {SummaryList} from "nhsuk-react-components"
import React from "react"
import {formatName, getAllAddressLines} from "../../../formatters/demographics"
import {newLineFormatter} from "../../common/newLineFormatter"

interface SummaryOrganization {
  name: string
  odsCode: string
  addressLines?: Array<string>
}

interface SummaryPractitionerRole {
  name: string
  professionalCodes: Array<string>
  telecom: string
  organization: SummaryOrganization
  parentOrganization: SummaryOrganization
}

const PractitionerRoleSummaryList = ({
  name,
  professionalCodes,
  telecom,
  organization,
  parentOrganization
}: SummaryPractitionerRole) => {
  const addressLineFragments = newLineFormatter(organization.addressLines)
  const professionalCodeFragments = newLineFormatter(professionalCodes)

  return (
    <SummaryList>
      <SummaryList.Row>
        <SummaryList.Key>Name</SummaryList.Key>
        <SummaryList.Value>{name}</SummaryList.Value>
      </SummaryList.Row>
      <SummaryList.Row>
        <SummaryList.Key>Professional Codes</SummaryList.Key>
        <SummaryList.Value>{professionalCodeFragments}</SummaryList.Value>
      </SummaryList.Row>
      <SummaryList.Row>
        <SummaryList.Key>Telecom</SummaryList.Key>
        <SummaryList.Value>{telecom}</SummaryList.Value>
      </SummaryList.Row>
      <SummaryList.Row>
        <SummaryList.Key>Organization</SummaryList.Key>
        <SummaryList.Value>{organization.name} ({organization.odsCode})</SummaryList.Value>
      </SummaryList.Row>
      <SummaryList.Row>
        <SummaryList.Key>Address</SummaryList.Key>
        <SummaryList.Value>{addressLineFragments}</SummaryList.Value>
      </SummaryList.Row>
      <SummaryList.Row>
        <SummaryList.Key>Trust / CCG</SummaryList.Key>
        <SummaryList.Value>{parentOrganization.name} ({parentOrganization.odsCode})</SummaryList.Value>
      </SummaryList.Row>
    </SummaryList>
  )
}

function createSummaryPractitionerRole(
  practitionerRole: PractitionerRole,
  practitioner: Practitioner,
  organization: Organization,
  healthcareService?: HealthcareService,
  location?: Location
): SummaryPractitionerRole {
  let organizationProps: SummaryOrganization
  let parentOrganizationProps: SummaryOrganization

  const professionalCodes = getProfessionalCodes([...practitionerRole.identifier, ...practitioner.identifier])

  if (healthcareService) {
    organizationProps = {
      name: healthcareService.name,
      odsCode: healthcareService.identifier[0].value,
      addressLines: getAllAddressLines(location.address)
    }
    parentOrganizationProps = {
      name: organization.name,
      odsCode: organization.identifier[0].value
    }
  } else {
    organizationProps = {
      name: organization.name,
      odsCode: organization.identifier[0].value,
      addressLines: getAllAddressLines(organization.address[0])
    }
    parentOrganizationProps = {
      name: organization.partOf.display,
      odsCode: organization.partOf.identifier.value
    }
  }

  return {
    name: formatName(practitioner.name[0]),
    professionalCodes: professionalCodes,
    telecom: practitionerRole.telecom[0].value,
    organization: organizationProps,
    parentOrganization: parentOrganizationProps
  }
}

function getProfessionalCodes(identifiers: Array<Identifier>): Array<string> {
  return identifiers.map(identifier => {
    switch (identifier.system) {
      case "https://fhir.nhs.uk/Id/sds-user-id":
        return "SDS User ID - " + identifier.value
      case "https://fhir.hl7.org.uk/Id/gmc-number":
        return "GMC Number - " + identifier.value
      case "https://fhir.hl7.org.uk/Id/gmp-number":
        return "GMP Number - " + identifier.value
      case "https://fhir.hl7.org.uk/Id/din-number":
        return "DIN Number - " + identifier.value
      case "https://fhir.hl7.org.uk/Id/gphc-number":
        return "GPHC Number - " + identifier.value
      case "https://fhir.hl7.org.uk/Id/hcpc-number":
        return "HCPC Number - " + identifier.value
      case "https://fhir.hl7.org.uk/Id/nmc-number":
        return "NMC Number - " + identifier.value
      case "https://fhir.hl7.org.uk/Id/nhsbsa-spurious-code":
        return "Spurious Code - " + identifier.value
      case "https://fhir.nhs.uk/Id/sds-role-profile-id":
        return "SDS Role Profile ID - " + identifier.value
      default:
        return "Unknown Code - " + identifier.value
    }
  })
}

export {
  SummaryPractitionerRole,
  PractitionerRoleSummaryList,
  createSummaryPractitionerRole
}
