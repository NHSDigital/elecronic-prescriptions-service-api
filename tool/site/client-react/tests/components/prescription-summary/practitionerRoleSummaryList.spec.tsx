import * as React from "react"
import PractitionerRoleSummaryList from "../../../src/components/prescription-summary/practitionerRoleSummaryList"
import pretty from "pretty"
import {summaryPractitionerRole} from "./props"
import {render} from "@testing-library/react"

test("Renders correctly", async () => {
  const component = <PractitionerRoleSummaryList {...summaryPractitionerRole}/>
  const {container} = render(component)
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})
