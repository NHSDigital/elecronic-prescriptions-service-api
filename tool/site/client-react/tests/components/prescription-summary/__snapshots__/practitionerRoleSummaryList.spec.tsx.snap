// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
"<dl class=\\"nhsuk-summary-list\\">
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Name</dt>
    <dd class=\\"nhsuk-summary-list__value\\">EDWARDS, Thomas (DR)</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Telecom</dt>
    <dd class=\\"nhsuk-summary-list__value\\">01234567890</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Organization</dt>
    <dd class=\\"nhsuk-summary-list__value\\">SOMERSET BOWEL CANCER SCREENING CENTRE (A99968)</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Address</dt>
    <dd class=\\"nhsuk-summary-list__value\\">MUSGROVE PARK HOSPITAL<br>TAUNTON<br>TA1 5DA</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Trust / CCG</dt>
    <dd class=\\"nhsuk-summary-list__value\\">TAUNTON AND SOMERSET NHS FOUNDATION TRUST (RBA)</dd>
  </div>
</dl>"
`;
